### PR TITLE
OpenStack: add Spec for HostedCluster and CAPO OpenStackCluster

### PR DESCRIPTION
**Pull request by MaysaMacedo**
/hold

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go [hunk 1]
```diff
@@ -39,31 +38,76 @@ func New(capiProviderImage string) *OpenStack {
...
+		for i := range openStackPlatform.Subnets {
+			openStackCluster.Spec.Subnets[i] = capo.SubnetParam{
+				ID: openStackPlatform.Subnets[i].ID,
+			}
+		}
+	} else {
+		machineNetworks := hcluster.Spec.Networking.MachineNetwork
+		openStackCluster.Spec.ManagedSubnets = make([]capo.SubnetSpec, len(machineNetworks))
+		// Only one Subnet is supported in CAPO
+		openStackCluster.Spec.ManagedSubnets[0] = capo.SubnetSpec{
```
Comment (MaysaMacedo) [inline hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go:79]: @EmilienM @mdbooth 
Any thoughts on the approach to handle the CAPO provisioned subnet?
Should we really consume the first CIDR from the hosted cluster and not define in the openstack platform spec? I'm enforcing the nameserver and allocationpool on that subnet.

File: hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go [hunk 2]
```diff
@@ -39,31 +38,76 @@ func New(capiProviderImage string) *OpenStack {
...
+	if openStackPlatform.NetworkMTU != nil {
+		openStackCluster.Spec.NetworkMTU = openStackPlatform.NetworkMTU
+	}
+	if openStackPlatform.ExternalNetwork != nil {
+		openStackCluster.Spec.ExternalNetwork = &capo.NetworkParam{ID: openStackPlatform.ExternalNetwork.ID}
+	}
+	if openStackPlatform.DisableExternalNetwork != nil {
+		openStackCluster.Spec.DisableExternalNetwork = openStackPlatform.DisableExternalNetwork
+	}
+	openStackCluster.Spec.ManagedSecurityGroups = &capo.ManagedSecurityGroups{}
```
Comment (MaysaMacedo) [inline hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go:109]: I didn't add additional SG rules as I'm not sure which additional ones might be required.
I thought about adding once we cover nodePools. Thoughts?
@EmilienM

**Comment by MaysaMacedo** [comment]
/hold

Let's hold merge until we discuss the topics I have

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 3]
```diff
@@ -1825,8 +1825,130 @@ type AzurePlatformSpec struct {
...
+	// DisableExternalNetwork specifies whether or not to attempt to connect the cluster
+	// to an external network. This allows for the creation of clusters when connecting
+	// to an external network is not possible or desirable, e.g. if using a provider network.
+	// +optional
+	DisableExternalNetwork *bool `json:"disableExternalNetwork,omitempty"`
+
+	// Tags to set on all resources in cluster which support tags
+	// +listType=set
+	// +optional
+	Tags []string `json:"tags,omitempty"`
```
Comment (mdbooth) [inline api/hypershift/v1beta1/hostedcluster_types.go:1900]: Does HyperShift already have a 'Tags' concept we could reuse here?

File: hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go (hunk 2 above)
Comment (EmilienM) [reply hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go:109]: we don't need SGs until we get workers deployed, so yeah we can wait.
Now, regarding the rules themselves, we'll want to compare what is managed by CAPO: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/pkg/cloud/services/networking/securitygroups_rules.go and what we need in OpenShift.

File: hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go [hunk 4]
```diff
@@ -39,31 +38,76 @@ func New(capiProviderImage string) *OpenStack {
...
 			Name:      hcluster.Name,
 			Namespace: controlPlaneNamespace,
 		},
 	}
-	openstackPlatform := hcluster.Spec.Platform.OpenStack
-	if openstackPlatform != nil {
-		openstackCluster.Spec.IdentityRef = capiopenstack.OpenStackIdentityReference{
-			Name:      openstackPlatform.CloudsYamlSecret.Name,
+	openStackPlatform := hcluster.Spec.Platform.OpenStack
+	if openStackPlatform != nil {
```
Comment (mdbooth) [inline hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go]: This logic is weird. We should flip this to `if openStackPlatform == nil` and return an error.

File: hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go (hunk 1 above)
Comment (mdbooth) [reply hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go:79]: Lets do that for now and ask the HCP folks what they think. We definitely don't want to require the user to correctly enter the same value twice in 2 different places.

File: hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go (hunk 2 above)
Comment (mdbooth) [reply hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go:109]: This is good for now.

**Review by mdbooth — APPROVED** [review]
/lgtm

File: hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go (hunk 4 above)
Comment (MaysaMacedo) [reply hypershift-operator/controllers/hostedcluster/internal/platform/openstack/openstack.go]: done

File: api/hypershift/v1beta1/hostedcluster_types.go (hunk 3 above)
Comment (MaysaMacedo) [reply api/hypershift/v1beta1/hostedcluster_types.go:1900]: Nop, this is specific per platform. For example aws accepts users tags.

**Pull request merged**